src/index.ts

```diff
@@ -21,38 +24,54 @@ function makeBase58CheckEncoder(p2pkhVersion: number, p2shVersion: number): (dat
         ) {
           throw Error('Unrecognised address format');
         }
-        addr = Buffer.concat([Buffer.from([p2pkhVersion]), data.slice(3, 3 + data.readUInt8(2))]);
+        if (p2pkhVersion[0] instanceof Array) {
+          addr = Buffer.concat([Buffer.from(p2pkhVersion[0]), data.slice(3, 3 + data.readUInt8(2))]);
+          return bs58Encode(addr);
+        }
+        addr = Buffer.concat([Buffer.from([p2pkhVersion[0]]), data.slice(3, 3 + data.readUInt8(2))]);
         // @ts-ignore
         return bs58Encode(addr);
       case 0xa9: // P2SH: OP_HASH160 <scriptHash> OP_EQUAL
         if (data.readUInt8(data.length - 1) !== 0x87) {
           throw Error('Unrecognised address format');
         }
-        addr = Buffer.concat([Buffer.from([p2shVersion]), data.slice(2, 2 + data.readUInt8(1))]);
+        if (p2shVersion[0] instanceof Array) {
+          addr = Buffer.concat([Buffer.from(p2shVersion[0].slice(0, 2)), data.slice(2, 2 + data.readUInt8(1))]);
+          return bs58Encode(addr);
+        }
+        addr = Buffer.concat([Buffer.from([p2shVersion[0]]), data.slice(2, 2 + data.readUInt8(1))]);
         return bs58Encode(addr);
       default:
         throw Error('Unrecognised address format');
     }
   };
 }
 
-function makeBase58CheckDecoder(p2pkhVersions: number[], p2shVersions: number[]): (data: string) => Buffer {
+function makeBase58CheckDecoder(
+  p2pkhVersions: (number | number[])[],
+  p2shVersions: (number | number[])[],
+): (data: string) => Buffer {
   return (data: string) => {
     const addr = bs58Decode(data);
-    const version = addr.readUInt8(0);
-    if (p2pkhVersions.includes(version)) {
-      return Buffer.concat([Buffer.from([0x76, 0xa9, 0x14]), addr.slice(1), Buffer.from([0x88, 0xac])]);
-    } else if (p2shVersions.includes(version)) {
-      return Buffer.concat([Buffer.from([0xa9, 0x14]), addr.slice(1), Buffer.from([0x87])]);
+    const version = addr.filter((b:any) => addr.slice(-20).indexOf(b) === -1);
```

Comment:
Can you document what's going on here in a comment? I can't follow it.

Reply:
Good point, I have no idea why `-20`. I only fixed the syntax error.